### PR TITLE
DBZ-693 Initial SQL statements

### PR DESCRIPTION
**Pull request by jpechane**
https://issues.jboss.org/browse/DBZ-693

### CODE REVIEW

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java [hunk 1]
```diff
@@ -508,6 +508,14 @@ public static EventProcessingFailureHandlingMode parse(String value) {
                                                          + "Each distinct MySQL installation should have a separate namespace and monitored by "
                                                          + "at most one Debezium connector.");
 
+    public static final Field ON_CONNECT_STATEMENTS = Field.create("database.initial.statements")
+                                                           .withDisplayName("Initial statements")
+                                                           .withType(Type.STRING)
+                                                           .withWidth(Width.LONG)
+                                                           .withImportance(Importance.LOW)
+                                                           .withDescription("A semicolon separated list of SQL statements to be executed when connection to database is established. "
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java]: I think we should clarify whether that only applies to snapshot reading, log reading, or both.

File: debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java [hunk 2]
```diff
@@ -586,6 +586,10 @@ public synchronized Connection connection() throws SQLException {
             if (conn == null) throw new SQLException("Unable to obtain a JDBC connection");
             // Always run the initial operations on this new connection
             if (initialOps != null) execute(initialOps);
+            final String statements = config.getString(JdbcConfiguration.ON_CONNECT_STATEMENTS);
+            if (statements != null) {
+                execute(statements.split(";"));
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java]: Should we handle the case where a `;` is given within a string literal?

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java (hunk 1 above)
Comment (gunnarmorling) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java]: Also the default (no statements) should be mentioned.

File: debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java (hunk 2 above)
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java]: TBH, I tried to postpone the scenario till it happens :-). But we can add a support for escaping - either using backslash or doubling the semicolon. In this case I'd probably prefer doubling the semicolon.

**Comment by jpechane** [comment]
@gunnarmorling Typos fixed, postgres support added, semicolon as a char supported

File: debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java [hunk 3]
```diff
@@ -581,11 +582,49 @@ public synchronized boolean isConnected() throws SQLException {
     }
 
     public synchronized Connection connection() throws SQLException {
+        return connection(true);
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java]: I think it'd be safer to not execute the statements when calling the parameterless `connection()`. Note that when using the PG connector, I see the insert being executed three times, so something is still not quite correct.

File: debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java [hunk 4]
```diff
@@ -581,11 +582,49 @@ public synchronized boolean isConnected() throws SQLException {
...
+    }
+
+    public synchronized Connection connection(boolean executeOnConnect) throws SQLException {
         if (conn == null) {
             conn = factory.connect(JdbcConfiguration.adapt(config));
             if (conn == null) throw new SQLException("Unable to obtain a JDBC connection");
             // Always run the initial operations on this new connection
             if (initialOps != null) execute(initialOps);
+            final String statements = config.getString(JdbcConfiguration.ON_CONNECT_STATEMENTS);
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java]: I'd suggest to extract all the parsing logic into its own method which then returns the list of statements to here.

File: debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java (hunk 3 above)
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java]: @gunnarmorling It is correct, as postgres connector creates multiple connections in sequence. The functionality is not for executing DML (albeit I abuse it for tests) just mainly for session configuration and in this case I expect the session to be configured every time the connection is created.

File: debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java (hunk 4 above)
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java]: @gunnarmorling Done

File: debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java (hunk 3 above)
Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java]: Ok, I'll expand the option's description a little bit then to clarify that fact.

**Comment by gunnarmorling** [comment]
Rebased and applied. Added one commit to expend the test a bit and update the connector option description. Thanks, @jpechane!

**Pull request closed without merging**